Add color space support for CanvasRenderingContext2D, ImageData, and ImageBitmap

ccameron-chromium opened the pull request:
Add color space support for CanvasRenderingContext2D, ImageData, and ImageBitmap.

NOTE: This PR is being uploaded for initial feedback, it is not ready for wider review (yet).

* [x] At least two implementers are interested (and none opposed):
  * Chromium
  * Mozilla
* [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
  * https://github.com/web-platform-tests/wpt/tree/master/html/canvas/element/wide-gamut-canvas
  * More being upstreamed
* [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
  * Chrome: https://crbug.com/1083693
  * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1703654
  * WebKit: https://bugs.webkit.org/show_bug.cgi?id=225140


(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/6562/acknowledgements.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/6562/ca7606c...5e497f3/acknowledgements.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">diff</a> )
<a href="https://whatpr.org/html/6562/canvas.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">/canvas.html</a>  ( <a href="https://whatpr.org/html/6562/ca7606c...5e497f3/canvas.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">diff</a> )
<a href="https://whatpr.org/html/6562/imagebitmap-and-animations.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">/imagebitmap-and-animations.html</a>  ( <a href="https://whatpr.org/html/6562/ca7606c...5e497f3/imagebitmap-and-animations.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">diff</a> )
<a href="https://whatpr.org/html/6562/index.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">/index.html</a>  ( <a href="https://whatpr.org/html/6562/ca7606c...5e497f3/index.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">diff</a> )
<a href="https://whatpr.org/html/6562/infrastructure.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/6562/ca7606c...5e497f3/infrastructure.html" title="Last updated on May 3, 2021, 10:21 PM UTC (5e497f3)">diff</a> )

kenrussell commented:
Note to reviewers: this work has been developed in https://github.com/WICG/canvas-color-space/blob/main/CanvasColorSpaceProposal.md and discussed at great length in the W3C's ColorWeb CG: https://github.com/w3c/ColorWeb-CG .

The artifacts here are the result of months of iteration and have been carefully considered.

domenic commented on source:
```diff
@@ -60070,11 +60070,14 @@ typedef (<span>HTMLOrSVGImageElement</span> or
          <span>ImageBitmap</span> or
          <span>OffscreenCanvas</span>) <dfn>CanvasImageSource</dfn>;
 
+enum <dfn>PredefinedColorSpace</dfn> { "<span data-x="dom-PredefinedColorSpace-srgb">srgb</span>", "<span data-x="dom-PredefinedColorSpace-display-p3">display-p3</span>" };
```
It seems like the web convention is "colorspace", one word, not "color space", two words: e.g. https://drafts.csswg.org/css-color-5/#colorspaces-section

This impacts the enum name (not observable) and the option names (observable).

domenic replied:
Hmm, https://drafts.csswg.org/css-color-4/ is less consistent. @tabatkins @svgeesus how should web APIs spell "color( )space"?

domenic replied:
At least one API already uses the two-word/camelcase version: https://html.spec.whatwg.org/#dom-imagebitmapoptions-colorspaceconversion

domenic replied:
I'm pretty sure this should be two words now (which also matches the existing spec), but I'd love if @tabatkins @svgeesus could confirm their plans on the CSS side.

Also we should update https://whatwg.org/style-guide#dictionary /cc @annevk

ccameron-chromium replied:
I've made sure this is two words throughout the document (except for this reference to the "Predefined colorspaces" section of CSS Color, for consistency with the target).

tabatkins replied:
Hm, the Color 4 spec is inconsistent, but definitely prefers "colorspace" (110 appearances) over "color space" (19 appearances). [Google Trends](https://trends.google.com/trends/explore?date=all&geo=US&q=%22color%20space%22,colorspace) has them fairly close, but with "color space" the clear winner.

Given that there's already an API spelling it as two words, I think it's clear we should continue with that, and I'll file a bug on the Color spec to get its spelling consistent.

svgeesus replied:
> 
> 
> Hmm, https://drafts.csswg.org/css-color-4/ is less consistent. @tabatkins @svgeesus how should web APIs spell "color( )space"?

I have tended to use colorspace but looking into technical books on color theory etc they all use **color space** so I closed [Tab's issue on naming](https://github.com/w3c/csswg-drafts/issues/6217) by doing the bulk edit to consistently say color space.

annevk replied:
I created a PR for the style guide at https://github.com/whatwg/whatwg.org/pull/359, I'll also resolve this.

domenic commented on source:
```diff
@@ -60070,11 +60070,14 @@ typedef (<span>HTMLOrSVGImageElement</span> or
 dictionary <dfn>CanvasRenderingContext2DSettings</dfn> {
   boolean <span data-x="dom-CanvasRenderingContext2DSettings-alpha">alpha</span> = true;
   boolean <span data-x="dom-CanvasRenderingContext2DSettings-desynchronized">desynchronized</span> = false;
+  <span>PredefinedColorSpace</span> <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</span> = "srgb";
```
Nit: cross-link this to the definition, `"<span data-x="dom-PredefinedColorSpace-srgb">srgb</span>"`

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60219,9 +60222,9 @@ interface mixin <dfn>CanvasDrawImage</dfn> {
 interface mixin <dfn>CanvasImageData</dfn> {
   // <span>pixel manipulation</span>
-  <span>ImageData</span> <span data-x="dom-context-2d-createImageData">createImageData</span>([EnforceRange] long sw, [EnforceRange] long sh);
+  <span>ImageData</span> <span data-x="dom-context-2d-createImageData">createImageData</span>([EnforceRange] long sw, [EnforceRange] long sh, optional <span>ImageDataSettings</span> settings);
```
```suggestion
  <span>ImageData</span> <span data-x="dom-context-2d-createImageData">createImageData</span>([EnforceRange] long sw, [EnforceRange] long sh, optional <span>ImageDataSettings</span> settings = {});
```

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60219,9 +60222,9 @@ interface mixin <dfn>CanvasDrawImage</dfn> {
+  <span>ImageData</span> <span data-x="dom-context-2d-createImageData">createImageData</span>([EnforceRange] long sw, [EnforceRange] long sh, optional <span>ImageDataSettings</span> settings);
   <span>ImageData</span> <span data-x="dom-context-2d-createImageData">createImageData</span>(<span>ImageData</span> imagedata);
-  <span>ImageData</span> <span data-x="dom-context-2d-getImageData">getImageData</span>([EnforceRange] long sx, [EnforceRange] long sy, [EnforceRange] long sw, [EnforceRange] long sh);
+  <span>ImageData</span> <span data-x="dom-context-2d-getImageData">getImageData</span>([EnforceRange] long sx, [EnforceRange] long sy, [EnforceRange] long sw, [EnforceRange] long sh, optional <span>ImageDataSettings</span> settings);
```
```suggestion
  <span>ImageData</span> <span data-x="dom-context-2d-getImageData">getImageData</span>([EnforceRange] long sx, [EnforceRange] long sy, [EnforceRange] long sw, [EnforceRange] long sh, optional <span>ImageDataSettings</span> settings = {});
```

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60297,15 +60300,20 @@ interface <dfn>TextMetrics</dfn> {
 };
 
+dictionary <dfn>ImageDataSettings</dfn> {
+  <span>PredefinedColorSpace</span> <span data-x="dom-ImageDataSettings-colorSpace">colorSpace</span> = "srgb";
```
Cross-link using `"<span data-x="dom-PredefinedColorSpace-srgb">srgb</span>"`

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60297,15 +60300,20 @@ interface <dfn>TextMetrics</dfn> {
 interface <dfn>ImageData</dfn> {
-  <span data-x="dom-imagedata">constructor</span>(unsigned long sw, unsigned long sh);
-  <span data-x="dom-imagedata">constructor</span>(<span data-x="idl-Uint8ClampedArray">Uint8ClampedArray</span> data, unsigned long sw, optional unsigned long sh);
+  <span data-x="dom-imagedata">constructor</span>(unsigned long sw, unsigned long sh, optional <span>ImageDataSettings</span> settings);
```
```suggestion
  <span data-x="dom-imagedata">constructor</span>(unsigned long sw, unsigned long sh, optional <span>ImageDataSettings</span> settings = {});
```

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60297,15 +60300,20 @@ interface <dfn>TextMetrics</dfn> {
-  <span data-x="dom-imagedata">constructor</span>(unsigned long sw, unsigned long sh);
-  <span data-x="dom-imagedata">constructor</span>(<span data-x="idl-Uint8ClampedArray">Uint8ClampedArray</span> data, unsigned long sw, optional unsigned long sh);
+  <span data-x="dom-imagedata">constructor</span>(unsigned long sw, unsigned long sh, optional <span>ImageDataSettings</span> settings);
+  <span data-x="dom-imagedata">constructor</span>(<span data-x="idl-Uint8ClampedArray">Uint8ClampedArray</span> data, unsigned long sw, optional unsigned long sh, optional <span>ImageDataSettings</span> settings);
```
```suggestion
  <span data-x="dom-imagedata">constructor</span>(<span data-x="idl-Uint8ClampedArray">Uint8ClampedArray</span> data, unsigned long sw, optional unsigned long sh, optional <span>ImageDataSettings</span> settings = {});
```

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -60339,6 +60347,9 @@ interface <dfn>Path2D</dfn> {
     data-x="concept-canvas-desynchronized">desynchronized</span>.</p>
+
+    <p>The <code data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</code> member
+    specifies the color space of the rendering context's <span>output bitmap</span>.
```
color space -> colorspace

domenic commented on source:
```diff
@@ -60413,6 +60424,12 @@ interface <dfn>Path2D</dfn> {
   where the latency between input and rasterization is critical.</p>
 
+  <p>The <code>CanvasRenderingContext2D</code> object also has a <dfn
+  data-x="concept-canvas-colorSpace">colorSpace</dfn> setting of type <code>PredefinedColorSpace</code>.
```
This concept also needs to be "colorspace" instead of colorSpace.

domenic commented on source:
```diff
@@ -60489,6 +60509,12 @@ interface <dfn>Path2D</dfn> {
    data-x="dom-CanvasRenderingContext2DSettings-desynchronized">desynchronized</code></dfn>.</p></li>
 
+   <li><p>If <var>settings</var>'s <dfn dict-member for="CanvasRenderingContext2DSettings"><code
+   data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</code></dfn> is a supported
```
Note the behavior here comes in two parts:

- Web IDL ensures that any non-`PredefinedColorSpace` values will throw a `TypeError`.
- Then, your "supported" check will take non-supported-but-implemented `PredefinedColorSpace` values and default them to srgb.

Is this intended? It makes sense if, e.g., some builds of a browser plan to only "support" srgb, even though the code is included which defines display-p3 as a member of the `PredefinedColorSpace` enum. In that case I'd rather phrase this as something like "If settings's colorSpace is supported by this user agent, then set..."

But if "supported" is supposed to mean "is a member of the implementation's `PredefinedColorSpace` definition", then this step doesn't work. Non-members will always throw a TypeError, per Web IDL rules.

ccameron-chromium replied:
My feeling is that the simplest behavior is to throw an exception if the user agent cannot create a canvas with the requested color space. This could be because the enum is invalid, or because of limitations of the user agent.

But this was something where we in the proposal's discussion wanted WhatWG's feedback as to what "fits best" with everything else.

The motivation for the behavior of "falling back gracefully to sRGB" is described in [this issue](https://github.com/WICG/canvas-color-space/blob/main/CanvasColorSpaceProposal.md#handling-unrecognized-color-spaces) in the proposal. This motivation is, as you point out, at odds with Web IDL. It is also at odds all other enum attributes (e.g, WebGLContextAttributes' powerPreference, ImageBitmapOptions colorSpaceConversion).

It may be that @junov or @fserb can give additional motivation for the fallback behavior.

On a related topic, the proposal includes [an interface](https://github.com/WICG/canvas-color-space/blob/main/CanvasColorSpaceProposal.md#color-spaces) for querying the available enum values, but I removed this from the PR based on feedback that that isn't a supported mechanism.

junov replied:
IMHO, the beauty of using a dictionary for the settings is that you can specify settings that are not supported by all browsers and it won't make the context creation fail, so there is no hard requirement upfront feature detection.  To preserve that philosophy, I think the color space field would need to be a string rather than an enum, as far as the dictionary's IDL is concerned.  But one could still assign PredefinedColorSpace values to that field if they prefer getting an exception rather than falling back.

domenic replied:
> But this was something where we in the proposal's discussion wanted WhatWG's feedback as to what "fits best" with everything else.

I think there are two behaviors that are a reasonable fit with how things are usually done in the spec:

- Use a string. Define the exact algorithm for converting the string to a color space (e.g. does `dIsPlay-P3` get converted to the display P3 colorspace, or to the default sRGB?). Ensure that `imageBitmap.colorSpace` and `getContextAttributes().colorSpace` returns the canonicalized name of the color space, not the input string.

  - The conversion algorithm converts the input string to the default if it is not supported by the browser, *or* if it's not in the list of predefined color spaces. You write out both of those things in the prose.

  - In this version the `enum` would not appear in the IDL at all.

- Use an enum. However, do not talk about "supported". Assume that an implementation which doesn't support, e.g. `"display-p3"` will just not include that in their `PredefinedColorSpace` enum, i.e. won't implement the full spec.

The current spec, which has two separate notions---"is in the enum" and "is supported"---is in a weird in-between place.

ccameron-chromium replied:
RE Enum vs String (prefer string, I'll update the PR to do this ASAP)
* CSS Color allows for "dIsPlay-P3". It would make sense to be consistent with that, which pushes towards DOMString.
* CSS Color also allows for custom profile color spaces. They are not part of this patch. Using a DOMString would leave the door open to allowing them.
* This is pushes me firmly in the direction of changing all of these to a DOMString. Would it be appropriate to write out the prose for converting the string to a color space here, or would it belong in the "Common microsyntaxes" section?

RE Error handling: If using a string, then we can choose any error handling mechanism we want (throw error or fall back). Let's circle back on this once the parsing syntax is added. Some thoughts are:
* Throwing an error is simpler, in my opinion
  * If the name is invalid, we give an error.
  * If the operation isn't supported by the implementation (e.g, there is a color space that an implementation can't render to for some reason), then we give an error.
  * In the event of an error, the caller is guaranteed that specifying "sRGB" won't fail.
* Falling back is a bit more complicated
  * But there is the "invalid values are ignored" idiom in places.
  * Should falling back apply to ImageData and ImageBitmap as well?
  * Note that ImageBitmap (being opaque) didn't get a colorSpace attribute, so there's no way for the caller to know if the colorSpace was ignored (unless we add an attribute, and do we want to?).

I'll update this to DOMString and we can then discuss the error situation (because at that point I think that will be the only remaining unresolved issue).

annevk replied:
I'm not a big fan of ASCII case-insensitive matching in APIs if we can avoid it.

Also, if you have a custom profile, presumably you need to describe it through various parameters? Wouldn't we overload this enum with a dictionary of sorts at that point?

ccameron-chromium replied:
Thanks for going over this with me.

> I'm not a big fan of ASCII case-insensitive matching in APIs if we can avoid it.

Indeed, we _can_ avoid it here. So I've removed that and stuck with the enum. I've also removed all of the "supported" text, because it's no longer relevant (if you've got the enum, you've got the functionality).

> Also, if you have a custom profile, presumably you need to describe it through various parameters? Wouldn't we overload this enum with a dictionary of sorts at that point?

For CSS profiles it's something like "--fogra39" in [this example](https://www.w3.org/TR/css-color-4/#cal-cmyk). But that may not be the right way to expose it to canvas (I could imagine creating a color space object, or dictionary, as you suggest). The overloading scheme seems like a likely solution.

domenic commented on source:
```diff
@@ -60540,6 +60566,22 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
 
   <hr>
 
+  <p>The <code>PredefinedColorSpace</code> enumeration is used to specify the <dfn>color space</dfn>
```
This `<dfn>` is not referenced. Maybe you meant to cross-reference `concept-canvas-colorSpace`?

ccameron-chromium replied:
Yes, that's correct. Changed to concept-canvas-colorSpace.

domenic commented on source:
```diff
@@ -60540,6 +60566,22 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
   <hr>
 
+  <p>The <code>PredefinedColorSpace</code> enumeration is used to specify the <dfn>color space</dfn>
+  of the canvas' backing store.</p>
```
```suggestion
  of the canvas's backing store.</p>
```

ccameron-chromium replied:
Updated here and the rest of the locations.

domenic commented on source:
```diff
@@ -60540,6 +60566,22 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
+  </p>
+
+  <p class="note">The color spaces' definitions, as well as functions for converting color spaces, are
+  found in the predefined color spaces section <ref spec=CSSCOLOR>.</p>
```
I'd suggest

```html
found in the <a href="https://drafts.csswg.org/css-color/#predefined">Predefined colorspaces</a> section of <cite>CSS Color</cite>. <ref spec=CSSCOLOR>
```

ccameron-chromium replied:
Thanks, that's what I wanted to convey. Updated to include this link and to clarify that conversion algorithms can be found at the link as well.

domenic commented on source:
```diff
@@ -63864,24 +63906,24 @@ try {
    pixels in the returned object are <span>transparent black</span>.</p></dd>
 
-   <dt><var>imagedata</var> = new <code data-x="dom-imagedata">ImageData</code>(<var>data</var>, <var>sw</var> [, <var>sh</var> ] )</dt>
+   <dt><var>imagedata</var> = new <code data-x="dom-imagedata">ImageData</code>(<var>data</var>, <var>sw</var> [, <var>sh</var>, <var>settings</var> ] )</dt>
```
Use nested brackets here to indicate optionality. (The syntax for these is very inconsistent; see https://github.com/whatwg/meta/issues/190. But that seems like a good local fix.)

ccameron-chromium replied:
Updated this to nest the brackets.

domenic commented on source:
```diff
@@ -63936,14 +63981,16 @@ try {
 
   <p>When the <code data-x="dom-imagedata">ImageData()</code> constructor is invoked with two
-  numeric arguments <var>sw</var> and <var>sh</var>, it must run these steps:</p>
+  numeric arguments <var>sw</var> and <var>sh</var> and an optional <span>ImageDataSettings</span>
```
```suggestion
  numeric arguments <var>sw</var> and <var>sh</var> and an optional <code>ImageDataSettings</code>
```

ccameron-chromium replied:
Updated.

domenic commented on source:
```diff
@@ -64029,12 +64078,16 @@ try {
+  values must not be <span data-x="concept-premultiplied-alpha">premultiplied by alpha</span>.
+  Pixel values must be converted from the <span>output bitmap</span>'s
+  <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">color space</span> to the returned
+  <code>ImageData</code> object's <span data-x="dom-imagedata-colorSpace">color space</span>.
```
Is this a well-known algorithm? Could we reference it? Or can we just assume all implementers will know how to do this?

ccameron-chromium replied:
Good point -- I added the word "using relative colorimetric intent", because that nails down the algorithm precisely.

The algorithm for relative colorimetric intent is outlined in:
https://www.w3.org/TR/css-color-4/#predefined-to-lab
https://www.w3.org/TR/css-color-4/#lab-to-predefined
I could include the following text, if it seems appropriate.
```
  The algorithm for this conversion is outlined in the
  <a href="https://drafts.csswg.org/css-color/#predefined">Predefined colorspaces</a> section of
  of <cite>CSS Color</cite>. <ref spec=CSSCOLOR>
```
I decided against including it here because it's already referenced above when defining the predefined color spaces (I added text to indicate that conversion algorithms are covered in the link).

domenic commented on source:
```diff
@@ -64065,6 +64118,15 @@ try {
+   <li><p>If <var>settings</var> is specified then let <var>cs</var> be the value of the
+   <dfn dict-member for="ImageDataSettings"><code data-x="dom-ImageDataSettings-colorSpace">colorSpace</code></dfn>
+   member of <var>settings</var>. If <var>settings</var> was not specified or if <var>cs</var> is not a supported
+   <code>PredefinedColorSpace</code> then let <var>cs</var> be "<code data-x="dom-PredefinedColorSpace-srgb">srgb</code>".</p>
```
Same question about "supported" here.

ccameron-chromium replied:
Updated the discussion at the thread at
https://github.com/whatwg/html/pull/6562#discussion_r611903604

ccameron-chromium replied:
Removed the "supported" text.

domenic commented on source:
```diff
@@ -65610,40 +65678,36 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  data-x="dom-context-2d-drawImage">drawImage()</code> method with an
-  <code>HTMLOrSVGImageElement</code> object), and when rendering the actual canvas bitmap to the
-  output device.</p>
+  <!--en-GB--><h5 id="color-space-conversion">Color spaces and color space conversion</h5>
```
Let's keep the old ID so as to not break links. You can do this by either leaving it as-is, or adding `<span id="oldid"></span>` after the opening `<h5>` tag.

ccameron-chromium replied:
Changed to the old ID.

domenic commented on source:
```diff
@@ -65610,40 +65678,36 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  <p class="note">Thus, in the 2D context, colors used to draw shapes onto the canvas will exactly
-  match colors obtained through the <code
-  data-x="dom-context-2d-getImageData">getImageData()</code> method.</p>
+  <p>The <code>canvas</code> APIs provide mechanisms for specifying the color space of the canvas'
```
```suggestion
  <p>The <code>canvas</code> APIs provide mechanisms for specifying the color space of the canvas's
```

here and throughout

ccameron-chromium replied:
Done.

domenic commented on source:
```diff
@@ -65610,40 +65678,36 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
   <div w-nodev>
 
-  <!--en-GB--><h5 id="colour-spaces-and-colour-correction">Color spaces and color correction</h5>
-
```
A significant amount of normative text here is deleted. Can you help reassure me that it's OK to remove these requirements?

ccameron-chromium replied:
Sure! Some of the text is moved around, too -- I'll just go through the points in the old code with an explanation of where they went.

* "The canvas APIs must perform color correction at only two points... when rendering images with their own..."
  * This is now covered by the paragraph at the bottom, "When drawing content to a 2D context, all inputs must be converted to the context's color space before drawing"
  * This is a bit more expansive, covering not just images, but all content (videos and CSS color spaces have color spaces too)
* "The canvas APIs must perform color correction at only two points... when rendering the actual canvas bitmap to the output device"
  * This is now covered by the phrase "Color space conversion must be applied to the canvas's backing store when rendering the canvas to the output device."
* "The toDataURL() method, when invoked..."
  * The new behavior is in the "Where output format allows it" paragraph
  * The new behavior is to always use a color profile, instead of never
* "In user agents that support CSS, the color space used by a <code>canvas</code> element must match the color space used for processing any colors for that element in CSS."
  * I think this was already half-obsolete -- the CSS colors now have a well-defined color space
  * That space is independent of the canvas's color space, and when and how conversion is applied is now defined
  * So I just deleted it
* "The gamma correction and color space information of images must be handled in such a way ..."
  * This is now at the top "this color space conversion must be identical to the color space conversion that would be applied to an <code>img</code> element..."
  * It also falls out of the fact that relative colorimetric rendering intent is path-independent (you can convert from space X to Y to Z to W or just X to W directly, and, ignoring quantization error, you get the same result)
  * This is also a bit shorter now because I chopped out gamma correction words (gamma is part of the color space, no need to mention is separately)
* "Furthermore, the rendering of images that have no color correction information..."
  * This behavior changed and is covered by the "Images that specify no color profile information" line
  * This matches what CSS Color Level 4 [already says](https://www.w3.org/TR/css-color-4/#untagged)

(Let me know if that clarifies everything!)

svgeesus replied:
> This is also a bit shorter now because I chopped out gamma correction words (gamma is part of the color space, no need to mention is separately)

Good. Partly because the transfer curve is part of the color space definition, and also because the term "gamma correction" is a poor one. Nothing is being corrected, or is less correct about linear; instead the term "gamma encoding" is better. Although still suboptimal because many transfer curves are not a simple power law (gamma is the exponent in a power law).

domenic commented on source:
```diff
@@ -95295,26 +95360,27 @@ dictionary <dfn>ImageBitmapOptions</dfn> {
      data-x="dom-ColorSpaceConversion-default">default</code></dfn>",
-     the color space conversion behavior is implementation-specific, and should be chosen according
-     to the color space that the implementation uses for drawing images onto the canvas.</p></li>
+     the color space conversion behavior is implementation-specific, and should be a conversion
```
I'm not sure this should be implementation-specific? Shouldn't the conversion be well-defined now that we have a clear target color space?

It seems like the previous spec was implementation-defined because "the color space that the implementation uses for drawing images onto the canvas" was not defined.

ccameron-chromium replied:
Yes, your suggestion makes sense.

I had left it as technically implementation-specific out of a sense of preserving backwards compatibility with whatever the historical meaning was (both webkit and chromium do a conversion to sRGB). But you're right, this should be nailed down.

I also specified that the conversion should be using relative colorimetric intent (as before), to be unambiguous about the conversion algorithm.

annevk replied:
If anything remains implementation-specific, we should use `<span>implementation-defined</span>` to denote it, by the way.

ccameron-chromium replied:
I've removed the implementation-specific parts and rephrased the "should" as "must".

domenic reviewed (commented):
Did a quick first pass; I've got to run to an appointment but this should help get started!

annevk commented on source:
```diff
@@ -60565,6 +60585,12 @@ interface <dfn>Path2D</dfn> {
    data-x="dom-CanvasRenderingContext2DSettings-desynchronized">desynchronized</code></dfn>.</p></li>
 
+   <li><p>If <var>settings</var>'s <dfn dict-member for="CanvasRenderingContext2DSettings"><code
+   data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</code></dfn> is a supported
```
What does supported mean? If it's not supported this will throw, no? (There's other instances of this too that are equally confusing to me.)

ccameron-chromium replied:
Updated the discussion at the thread at
https://github.com/whatwg/html/pull/6562#discussion_r611903604

ccameron-chromium replied:
Removed the "supported" text (if the color space is a valid enum value, it will work).

annevk commented on source:
```diff
@@ -60616,6 +60642,23 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
+
+  <p>The "<dfn enum-value for="PredefinedColorSpace"><code
+  data-x="dom-PredefinedColorSpace-srgb">srgb</code></dfn>" value indicates the sRGB color space.
+  </p>
```
No newline right before a closing tag. (Happens elsewhere too.)

ccameron-chromium replied:
Done (I think I caught all of the instances of this).

annevk commented on source:
```diff
@@ -60616,6 +60642,23 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
+
+  <p class="note">The color spaces' definitions, as well as algorithms for converting between color
+  spaces, are found in the <a href="https://drafts.csswg.org/css-color/#predefined">Predefined
+  colorspaces</a> section of <cite>CSS Color</cite>. <ref spec=CSSCOLOR>
```
Missing closing tag.

ccameron-chromium replied:
Fixed!

annevk commented on source:
```diff
@@ -65686,40 +65756,36 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  <p>In user agents that support CSS, the color space used by a <code>canvas</code> element must
-  match the color space used for processing any colors for that element in CSS.</p>
+  <p>Images that specify no color profile information are assumed to be in the sRGB color space, and
+  must be rendered using the same color space conversion as is applied to sRGB images.</p>
```
Is this redundant with CSS? If so, might be better as a note.

ccameron-chromium replied:
Yes, it is redundant with CSS. Changed to a note that references the CSS spec.

annevk commented on source:
```diff
@@ -65686,40 +65756,36 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  of images that have no color correction information (such as those returned by the <code
-  data-x="dom-canvas-toDataURL">toDataURL()</code> method) must be rendered with no color
-  correction.</p>
+  <p>Where output format allows it, the <code data-x="dom-canvas-toDataURL">toDataURL()</code> method, when
```
output formats allow*

It seems better to define this as part of the `toDataURL()` definition though. Action at a distance is best avoided in specifications.

ccameron-chromium replied:
Good point. Added this to the "Serializing bitmaps to a file" section. I also clarified that for formats that do not support embedded color profiles, the bitmap should be converted to sRGB before being serialized (the goal being to have a subsequent drawImage of the result from toDataURL be as close to a no-op as possible).

annevk reviewed (commented):
Very exciting to see some incremental progress here!

@whatwg/canvas might want to review this as well.

ccameron-chromium commented on source:
```diff
@@ -60146,11 +60146,14 @@ typedef (<span>HTMLOrSVGImageElement</span> or
 dictionary <dfn>CanvasRenderingContext2DSettings</dfn> {
   boolean <span data-x="dom-CanvasRenderingContext2DSettings-alpha">alpha</span> = true;
   boolean <span data-x="dom-CanvasRenderingContext2DSettings-desynchronized">desynchronized</span> = false;
+  <span>PredefinedColorSpace</span> <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</span> = "srgb";
```
Yes, updated.

ccameron-chromium commented on source:
```diff
@@ -60373,15 +60376,20 @@ interface <dfn>TextMetrics</dfn> {
 };
 
+dictionary <dfn>ImageDataSettings</dfn> {
+  <span>PredefinedColorSpace</span> <span data-x="dom-PredefinedColorSpace-srgb">colorSpace</span> = "srgb";
```
Updated.

ccameron-chromium commented on source:
```diff
@@ -60489,6 +60500,12 @@ interface <dfn>Path2D</dfn> {
   where the latency between input and rasterization is critical.</p>
 
+  <p>The <code>CanvasRenderingContext2D</code> object also has a <dfn
+  data-x="concept-canvas-colorSpace">colorSpace</dfn> setting of type <code>PredefinedColorSpace</code>.
```
Changed to concept-canvas-color-space.

annevk replied:
You haven't changed the contents to "color space". They are still "colorSpace" as far as I can tell.

ccameron-chromium replied:
Oops, I only changed the concept-canvas-color-space! Fixed this to remove the uses of colorSpace (except for places where it is referring to a specific key).

ccameron-chromium commented:
Thanks everyone for the comments so far. I think that I've covered all of the issues, and I think we've landed in a good place with respect to the questions that we came in here with.

Let me know if there are any other issues or concerns! (And if I've missed any of the corrections -- I did proofread it again again but ... I have an ECC module when reading text that I wrote).

annevk commented on source:
```diff
@@ -60500,6 +60518,9 @@ interface <dfn>Path2D</dfn> {
    <li><p><code data-x="concept-canvas-desynchronized">desynchronized</code>, set to this context's
    <span data-x="concept-canvas-desynchronized">desynchronized</span>.</p></li>
+
+   <li><p><code data-x="concept-canvas-color-space">colorSpace</code>, set to this context's
```
color space

domenic replied:
The problem here is the first one (on this line) should be `dom-canvasrenderingcontext2dsettings-colorspace`

ccameron-chromium replied:
Ah, all bullet points in this section make that mistake! I updated concept-canvas-alpha,desynchronized,color-space to dom-canvasrenderingcontext2dsettings-alpha,desynchronized,colorSpace

annevk commented on source:
```diff
@@ -60500,6 +60518,9 @@ interface <dfn>Path2D</dfn> {
    <span data-x="concept-canvas-desynchronized">desynchronized</span>.</p></li>
+
+   <li><p><code data-x="concept-canvas-color-space">colorSpace</code>, set to this context's
+   <span>output bitmap</span>'s <span data-x="concept-canvas-color-space">color space</span>.</p></li>
```
It's not associated with the output bitmap though, right? It's part of the context and affects how the output bitmap is rendered. Otherwise you need a different concept- here.

ccameron-chromium replied:
Conceptually there are three distinct things that I may not have been precise about in my language. 
* canvas element, which doesn't have its own color space
* rendering context (2D or WebGL), which has a color space that affects how things are rendered
* output bitmap (or backing store), which will be a bitmap where the pixels are in the rendering context's color space

I did a quick sweep of the patch and updated "output bitmap" to be "context" here. Later on, in the "When drawing content to a 2D context..." text, I referred to the "canvas's color space" when the "context's color space" would have been more precise.

annevk commented on source:
```diff
@@ -64105,13 +64153,17 @@ try {
-  values must not be <span data-x="concept-premultiplied-alpha">premultiplied by alpha</span>.</p>
+  values must not be <span data-x="concept-premultiplied-alpha">premultiplied by alpha</span>.
+  Pixel values must be converted from the <span>output bitmap</span>'s
+  <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">color space</span> to the returned
```
This looks wrong as well.

ccameron-chromium replied:
Gah! Sorry missing it! This time did the s/dom-CanvasRenderingContext2DSettings-colorSpace/concept-canvas-color-space/ for real.

(Note that I didn't remove the "output bitmap" part, because "bitmap" seems most correct in this situation, although equivalent to the "context").

annevk commented on source:
```diff
@@ -64105,13 +64153,17 @@ try {
+  values must not be <span data-x="concept-premultiplied-alpha">premultiplied by alpha</span>.
+  Pixel values must be converted from the <span>output bitmap</span>'s
+  <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">color space</span> to the returned
+  <code>ImageData</code> object's <span data-x="dom-imagedata-colorSpace">color space</span> using
```
This too. This is referencing an IDL attribute and those should use `<code>` (and camelcase).

ccameron-chromium replied:
Changed the first one to s/settings-colorSpace/concept-color-space/, changed the second on to s/span/code/

annevk commented on source:
```diff
@@ -65686,39 +65752,27 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  match the color space used for processing any colors for that element in CSS.</p>
+  <p class="note">The <a href="https://drafts.csswg.org/css-color/#untagged">Color Spaces of Untagged Colors</a>
+  section of <cite>CSS Color</cite> specifies that images that specify no color profile information are assumed
+  to be in the sRGB color space.<ref spec=CSSCOLOR></p>
```
Maybe merge this note into the note below?

ccameron-chromium replied:
Makes sense -- merged them into a single note.

annevk reviewed (commented):
I found a couple cases remaining in the color space vs colorSpace category.

domenic commented on source:
```diff
@@ -65686,39 +65752,27 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
+  <p>Color space conversion must be applied to the canvas's backing store when rendering the canvas to the
+  output device. This color space conversion must be identical to the color space conversion that would
+  be applied to an <code>img</code> element with a color profile that specifies the same color space as
+  the canvas's backing store.</p>
```
"the same color space as the canvas's concept-canvas-color-space", maybe? (Hyperlinked)

ccameron-chromium replied:
Added the link, but left it reading ```the same <span data-x="concept-canvas-color-space">color space</span> as the canvas's backing store.``` Because we're talking about the pixels in the buffer, I think that talking about the color space of the backing store is the right scope (even though the "backing store" space and "context" space are the same).

domenic commented on source:
```diff
@@ -65686,39 +65752,27 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  of images that have no color correction information (such as those returned by the <code
-  data-x="dom-canvas-toDataURL">toDataURL()</code> method) must be rendered with no color
-  correction.</p>
+  <p>When drawing content to a 2D context, all inputs must be converted to the context's color space
```
"the canvas's concept-canvas-color-space" here and below?

ccameron-chromium replied:
Changed to ``converted to the context's <span data-x="concept-canvas-color-space">color space</span>``` in the three references.

domenic commented on source:
```diff
@@ -65686,39 +65752,27 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  data-x="dom-context-2d-drawImage">drawImage()</code> method to render the output of the <code
-  data-x="dom-canvas-toDataURL">toDataURL()</code> method to the canvas, given the appropriate
-  dimensions, has no visible effect.</p>
+  <p class="note">The color space for CSS colors is defined in <cite>CSS Color</cite>: <ref
+  spec=CSSCOLOR>. There do not exist any inputs to a 2D context for which the color space is undefined.</p>
```
```suggestion
  <p class="note">The color space for CSS colors is defined in <cite>CSS Color</cite>. There do not
  exist any inputs to a 2D context for which the color space is undefined. <ref spec=CSSCOLOR></p>
```

(stylistic nit)

ccameron-chromium replied:
Chnaged to a terminal ```<ref spec=CSSCOLOR></p>``` in the merged comment.

domenic commented on source:
```diff
@@ -64105,13 +64153,17 @@ try {
+  Pixel values must be converted from the <span>output bitmap</span>'s
+  <span data-x="dom-CanvasRenderingContext2DSettings-colorSpace">color space</span> to the returned
+  <code>ImageData</code> object's <span data-x="dom-imagedata-colorSpace">color space</span> using
+  relative colorimetric intent.
```
Can "relative colorimetric intent" hyperlink anywhere? (Fine if not, but it would be a nice bonus if it could.)

svgeesus replied:
How about https://drafts.csswg.org/css-color-4/#valdef-color-profile-rendering-intent-relative-colorimetric in CSS Color 4

ccameron-chromium replied:
Added the link to the places where it was referenced.

domenic reviewed (commented):
Looking great; my review, like @annevk's, is down to some nits at this point.

annevk commented on source:
```diff
@@ -60477,6 +60485,10 @@ interface <dfn>Path2D</dfn> {
+
+    <p>The <code data-x="dom-CanvasRenderingContext2DSettings-colorSpace">colorSpace</code> member
+    specifies the <span data-x="concept-canvas-color-space">color space</span> of the rendering
+    context's <span>output bitmap</span>.
```
I think as per the definition below the color space is of the rendering context and used for the rendering context's output bitmap, no? Perhaps it's okay to leave this as-is though as properly fixing this is a lot more work.

ccameron-chromium replied:
I agree we should change this to just "color space of the rendering context.", so I've made that change.

Indeed we may want to make another pass to clarify usage of "output bitmap" across the board.

annevk commented on source:
```diff
@@ -64203,6 +64256,14 @@ try {
+   "<code data-x="dom-PredefinedColorSpace-srgb">srgb</code>".</p>
+
+   <p>Initialize the <dfn attribute for="ImageData"><code data-x="dom-imagedata-colorSpace">colorSpace</code></dfn>
+   attribute of <var>imageData</var> to <var>cs</var>.</p></li>
```
This is missing an `<li>`, but this and the previous step can be simplified as the dictionary is always passed from the perspective of this algorithm (due to IDL) so you don't have to specify the defaulting here again. You can just say 'Initialize ... to _settings_["colorSpace"].'

ccameron-chromium replied:
Fixed the ```<li>``` part.

Holding off on the rest of the change because of the below question about optional-izing the colorSpace key.

annevk commented on source:
```diff
@@ -64167,13 +64215,18 @@ try {
-  steps:</p>
+  <var>pixelsPerRow</var>, an optional <code
+  data-x="idl-Uint8ClampedArray">Uint8ClampedArray</code> <var>source</var>, and an optional
+  <code>ImageDataSettings</code> <var>settings</var>, it must run these steps:</p>
```
This argument isn't optional. It's always given due to IDL.

domenic replied:
I think for this "create an ImageData object" algorithm it might be optional?

annevk replied:
I guess that is correct, although that did make me notice that in one instance of createImageData() where it is not optional, it's not defined to be passed on.

And in the description of getImageData() it's treated as optional but it's not.

But yeah, there is one instance of createImageData() that does not pass it and it's probably better to keep the defaulting in the "create an ImageData object" algorithm.

ccameron-chromium replied:
> I guess that is correct, although that did make me notice that in one instance of createImageData() where it is not optional, it's not defined to be passed on.

Is this the "When the createImageData() method is invoked with a single imagedata argument" instance? The wording there could be re-worked if you'd like.

> And in the description of getImageData() it's treated as optional but it's not.

Is this the "it must create an ImageData object, with parameter pixelsPerRow set to sw, parameter rows set to sh, and parameter settings set to settings if specified" text? And the fix is to remove the "if specified" at the end?

IMO that paragraph is getting too long and should be changed to a list of steps.

> But yeah, there is one instance of createImageData() that does not pass it and it's probably better to keep the defaulting in the "create an ImageData object" algorithm.

While we're on this topic, I'd like the WhatWG reviewers to take a moment to scrutinize the proposal's behavior for getImageData. This was something that we felt ambivalent about in the proposal, and would appreciate WhatWG's thoughts.

* The behavior we landed on was to convert the data to sRGB unless another color space is explicitly specified.
  * This happens by ImageDataSettings.colorSpace defaulting to sRGB.
  * The downside is that a default getImageData-then-putImageData can lose information
* The alternative would be to return data in the color space of the canvas
  * This would require having ImageDataSettings.colorSpace not have a default value
    * And then adding some language about what how to handle that
  * It would bring up the question of what createImageData should do by default
    * Should it match getImageData and use the context's color space?
    * Should it match the ImageData constructors that default to sRGB

annevk replied:
In order:

1. I meant the other case as when you invoke it with an ImageData object there's no settings.
2. Correct.
3. I'd be curious to learn why you landed on that behavior. If you asked me right now, I'd be inclined to say that `getImageData()`-then-`putImageData()` should not lose information (though there might be subpixel cases where they nonetheless would?); and that `createImageData()` (when not passed `ImageData`) should use the color space from the rendering context (unless overridden). When it is passed `ImageData` it should just copy from the passed in instance as it does now, but also copy the color space.

ccameron-chromium replied:
Thanks! I've updated the PR to hopefully cover all of the issues.

1. Updated, using the below defaultColorSpace
2. Updated, and changed it to a step-by-step section
3. This was for backwards compatibility. I had been the main advocate of this position
  * Backwards compatibility in the sense that "getImageData has always returned sRGB data"
  * But I can see that there is an equal if not more compelling case that ```getImageData```-then-```putImageData``` should not lose information
  * I've updated the ```createImageData```s and ```getImageData``` to behave the way you suggested here by adding an optional defaultColorSpace parameter to the ImageData creation algorithm.

jdashg commented on source:
```diff
@@ -64090,9 +64136,10 @@ try {
-  <var>sh</var>, it must run these steps:</p>
+  argument being an <code data-x="idl-Uint8ClampedArray">Uint8ClampedArray</code> <var>source</var>,
+  its second and optional third arguments being numeric arguments <var>sw</var> and
+  <var>sh</var>, and fourth <code>ImageDataSettings</code> argument <var>settings</var>, it
```
Isn't this the 'optional fourth [...] argument'?

ccameron-chromium replied:
This was based on the feedback in [this thread](https://github.com/whatwg/html/pull/6562#discussion_r615706541).

The IDL now reads
```
    constructor(Uint8ClampedArray data,
                unsigned long sw,
                optional unsigned long sh,
                optional ImageDataSettings settings = {});
```
IIUC, because a default value for settings is given, it's not "optional" from the perspective of the implementation.

Let me know if I didn't land in the right place here.

jdashg commented on source:
```diff
@@ -64142,38 +64193,55 @@ try {
+  <p>When the <dfn method for="CanvasImageData"><code
+  data-x="dom-context-2d-getImageData">getImageData()</code></dfn> method is invoked with four
+  numerical arguments <var>sx</var>, <var>sy</var>, <var>sw</var>, and <var>sh</var>, and a
+  fifth <code>ImageDataSettings</code> argument <var>settings</var>, it must run these steps:</p>
```
s/a fifth/an optional fifth/?

ccameron-chromium replied:
(same thing about user-optional vs implementation-optional)

annevk reviewed (commented):
I found some more issues, some might be substantive. (I wish I had gotten further with #3424 as a lot of this would be helped with having more modern foundations.)

ccameron-chromium commented on source:
```diff
@@ -64129,10 +64177,13 @@ try {
-  <code>ImageData</code> object</span>, with parameter <var>pixelsPerRow</var> set to the
-  absolute magnitude of <var>sw</var>, and parameter <var>rows</var> set to the absolute magnitude
-  of <var>sh</var>. Initialize the image data of the new <code>ImageData</code> object to
+  invoked with two numeric arguments <var>sw</var> and <var>sh</var> and an <code>ImageDataSettings</code>
```
Fixed.

ccameron-chromium commented on source:
```diff
@@ -64142,38 +64193,55 @@ try {
+  fifth <code>ImageDataSettings</code> argument <var>settings</var>, it must run these steps:</p>
+
+  <ol>
+    <li><p>If either the <var>sw</var> or <var>sh</var> arguments are zero, then throw an
```
Fixed.

ccameron-chromium commented on source:
```diff
@@ -64142,38 +64193,55 @@ try {
+    <li><p>If either the <var>sw</var> or <var>sh</var> arguments are zero, then throw an
+    <span>"<code>IndexSizeError</code>"</span> <code>DOMException</code>.</p></li>
+
+    <!--ADD-TOPIC:Security-->
```
I'd prefer to leave it there and we can remove it separately later (since that change wouldn't be related to the core of the patch).

ccameron-chromium commented on source:
```diff
@@ -60678,6 +60703,22 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
+
+  <p class="note">The color spaces' definitions, as well as algorithms for converting between color
+  spaces, are found in the <a href="https://drafts.csswg.org/css-color/#predefined">Predefined
+  colorspaces</a> section of <cite>CSS Color</cite>. <ref spec=CSSCOLOR></p>
```
Fixed (the target text had changed!).

ccameron-chromium commented on source:
```diff
@@ -64203,6 +64271,17 @@ try {
+   <var>imageData</var> to the <dfn dict-member for="ImageDataSettings"><code data-x="dom-ImageDataSettings-colorSpace">colorSpace</code></dfn>
+   member of <var>settings</var>.</p>
+
+   <p>Otherwise, if <var>defaultColorSpace</var> is specified, then initialize the
```
Fixed.

ccameron-chromium commented on source:
```diff
@@ -95092,6 +95177,7 @@ dictionary <dfn>ImageBitmapOptions</dfn> {
   <span>ImageOrientation</span> <span data-x="dom-ImageBitmapOptions-imageOrientation">imageOrientation</span> = "<span data-x="dom-ImageOrientation-none">none</span>";
   <span>PremultiplyAlpha</span> <span data-x="dom-ImageBitmapOptions-premultiplyAlpha">premultiplyAlpha</span> = "<span data-x="dom-PremultiplyAlpha-default">default</span>";
   <span>ColorSpaceConversion</span> <span data-x="dom-ImageBitmapOptions-colorSpaceConversion">colorSpaceConversion</span> = "<span data-x="dom-ColorSpaceConversion-default">default</span>";
+  <span>PredefinedColorSpace</span> <span data-x="dom-ImageBitmapOptions-colorSpace">colorSpace</span> = "<span data-x="dom-PredefinedColorSpace-srgb">srgb</span>";
```
I think it's probably best for me to take the `createImageBitmap` part out of this PR, and then take up the issue along with (or on top of) the changes you were suggesting for createImageBitmap.

The original version of the patch had a much shorter section on `createImageBitmap` and left things more vague and implementation-defined. I've reverted back to an even smaller version of that (not adding the `colorSpace` field, and just a few wording here and there).

It was for WebGL and WebGPU textures that we added the `colorSpace` field, but the same behavior can be accomplished in other ways, and it seems better to leave it out.

> While reading through the new `createImageBitmap()` text I started to wonder if these options should also apply to `<canvas>` and `ImageData` now that they have color space information that can differ from the desired color space. (If nobody wants to implement that I think at the very least we should throw.)

It depends on how the ImageBitmap is going to be used.

If the goal is to use the ImageBitmap for a WebGL or WebGPU texture, then _all_ content should be converted to a very well-defined pixel format and color space. Because WebGL and WebGPU aren't color managed, the application needs to know its exact starting place, to do the right math for their manual color management. For this use case, `createImageBitmap` should arguably allow specifying a pixel format, too (RGBA v BGRA v R10G10B10A2 v half-float, etc).

If the goal is to use the ImageBitmap in a 2D or ImageBitmap RenderingContext, then the exact internal representation doesn't matter, and _no_ content should have its pixel format or color space altered. The 2D and ImageBitmap contexts do color management and pixel format conversion internally. Ideally, from a performance perspective, the content shouldn't be converted at all, and `createImageBitmap` would create some sort of copy-on-write alias of the original source content.

> Also, should concept-imagebitmap-bitmap-data be updated to also have a color space, apart from a width and height?

I don't think so. In the case where we're not sending the data into WebGL or WebGPU, we want ImageBitmaps to be in whatever internal color space is most appropriate for them (which for many elements may even be some YUV video format). Any color space we would expose would likely be a lie in some circumstances.

kenrussell replied:
It would be great if we could press ahead and try to get the `colorSpace` dictionary member specified here instead of pushing it off. ImageBitmap is the path recommended to web developers nowadays for uploading images into WebGL and WebGPU, since it performs all conversions early on, and avoids synchronous re-conversions at the time of use. ImageBitmap behaves distinctly from `<canvas>` and `ImageData` in this regard. WebGL, for example, will perform certain operations like vertical flips for most image sources, but by design, does *not* do these for ImageBitmaps.

ccameron-chromium replied:
Given that ImageBitmap is already being torn between two separate uses (WebGL/GPU vs 2D/ImageBitmap), I don't think it's appropriate to add features to it in this PR. It needs careful separate consideration, and a clear vision of its future.

jdashg replied:
@kenrussell ImageBitmap is recommended in Chrome, but not in Firefox. Usually you're better off without ImageBitmap intermediaries in Firefox.

kenrussell replied:
If Firefox implemented all of ImageBitmap's creation options, it would be a more performant path than uploading HTMLImageElements to WebGL textures. Three.js, the most popular WebGL library, switched its preferred texture upload path to use ImageBitmap years ago, reducing dropped frames in applications.

This having been said, I now understand better the desire to upgrade ImageData and CanvasRenderingContext2D separately, so OK, fine by me to postpone adding color space control to ImageBitmap.

jdashg replied:
> If Firefox implemented all of ImageBitmap's creation options, it would be a more performant path than uploading HTMLImageElements to WebGL textures. 

I don't necessarily agree with this. Happy to chat elsewhere, but let's not overgeneralize.

annevk replied:
If I understand this correctly, `ImageBitmap` can be passed directly to WebGL and this is a thing that some applications do. I think that means that we have to consider `ImageBitmap` as part of this PR. Otherwise we leave it undefined what happens when you create an `ImageBitmap` from a non-sRGB `CanvasRenderingContext2D` or `ImageData` and have to effectively settle for that what the majority of applications will come to depend upon.

Or am I missing something here?

ccameron-chromium replied:
I would strongly prefer to cover `ImageBitmap`/WebGL issues in a separate PR.

There is separate work in Khronos for WebGL (which I believe jdashg volunteered for), and I feel that this will significantly expand the scope of the patch (which is narrowly focused on 2D/ImageData). The class of under-defined behavior that you mentioned already exists and is very wide (importing video to WebGL is the most prevalent example), and I think it needs its own focused treatment, rather than to be being picked up incidentally.

ccameron-chromium replied:
Oh, since it wasn't clear, I'm volunteering to start on that cleanup starting immediately, I just would prefer it to be separated out from this PR. I can also incorporate [#3424](https://github.com/whatwg/html/pull/3424) into that work if you'd like.

annevk replied:
My worry is that we end up in an inconsistent state, which is generally something we try to avoid. I'll defer to @domenic as to whether that's blocking here.

ccameron-chromium replied:
Thanks! I think we'll be able to resolve the issue on the WebGL side of things (and if that goes as I hope, we'll be able to just tighten up some language on this side).

domenic replied:
I took a look and I'm comfortable with how this only touches one part of the problem for now, cleanly leaving the rest of future work. I trust @ccameron-chromium to continue with that work.

I'm not really able to evaluate the extent of the problem, but I found

> The class of under-defined behavior that you mentioned already exists and is very wide (importing video to WebGL is the most prevalent example), and I think it needs its own focused treatment, rather than to be being picked up incidentally.

compelling to the extent that it's true, and if we can solve it all at once that'd be lovely.

ccameron-chromium commented on source:
```diff
@@ -64002,24 +64043,24 @@ try {
 
   <dl class="domintro">
    <dt><var>imagedata</var> = new <code subdfn data-x="dom-imagedata">ImageData</code>(<var>sw</var>, <var>sh</var>)</dt>
```
Good point -- I separated them out. I also moved the two `createImageData` functions next to each other (they were shuffled because of the shared definitions).

domenic commented:
I'm working on an editorial update to modernize some stuff while I'm in the area, but it looks good so far. I'll push soon.

However I noticed one potential bigger problem. OffscreenCanvasRenderingContext2D also mixes in CanvasImageData, but does not have a color space member, so the definition of createImageData() doesn't work.

ccameron-chromium commented:
> However I noticed one potential bigger problem. OffscreenCanvasRenderingContext2D also mixes in CanvasImageData, but does not have a color space member, so the definition of createImageData() doesn't work.

I had thought that it mixed in the settings parsing and the color space member, but you're right, it doesn't! I've added these to OffscreenCanvasRenderingContext2D now.

I noticed that OffscreenCanvasRenderingContext2D has a similar alpha property which can be set via CanvasRenderingContext2DSettings. What's odd is that the spec refers to being able to dynamically change it, but there's no API for that (and the setting isn't even accessible -- it's not an attribute, and there is no getContextAttributes mixed in). Is that a known issue?

domenic commented:
> I noticed that OffscreenCanvasRenderingContext2D has a similar alpha property which can be set via CanvasRenderingContext2DSettings. What's odd is that the spec refers to being able to dynamically change it, but there's no API for that (and the setting isn't even accessible -- it's not an attribute, and there is no getContextAttributes mixed in). Is that a known issue?

This is probably worth tracking separately... I agree as written it's odd. I think it's trying to say "Set it to true by default. The 'offscreen 2D context creation algorithm' might change this."

domenic reviewed (approved):
I'm happy with this PR.

I pushed another commit which does a pretty large editorial modernization of the touched sections as they were starting to get really crufty and I was sad. That sort of thing shouldn't be on contributors but while we're in the area I like to leave things better than they started.

Can you make sure to file a WebKit bug and update the OP with it? Also can we get a commitment to test all the things that changed during review, e.g. handling of invalid values, getImageData-then-putImageData, ... any others I missed?

@annevk are you happy with this landing after those steps?

annevk reviewed (commented):
I found a couple nits that I can address before landing. I agree with @domenic on the other remaining TODOs here.

One thing I would really appreciate (and maybe we should add this to the specification as well) is a concrete example. E.g., if you create a context using 'display-p3' and then use `fillStyle = hexNotation` (as per https://drafts.csswg.org/css-color/#hex-notation). Presumably if you then do `getImageData()` (and don't do color space conversion; but I guess also if you do) you would see some different values there in the data. I think that would be helpful to demonstrate.

ccameron-chromium commented on source:
```diff
@@ -64001,25 +64043,20 @@ try {
+   <dt><var>imagedata</var> = new <code subdfn data-x="dom-imagedata">ImageData</code>(<var>sw</var>, <var>sh</var> [, <var>settings</var>])</dt>
    <dd>
-    <p>Returns an <code>ImageData</code> object with the given dimensions. All the pixels in the
+    <p>Returns an <code>ImageData</code> object with the given dimensions and the color space indicated by <var>settings</var>. All the pixels in the
```
I've gone through the patch and quickly updated all touched lines to wrap 100, with the exception of some URLs and dt tags that look as though they are conventionally exceptions.

ccameron-chromium commented on source:
```diff
@@ -64029,10 +64066,26 @@ try {
 
-   <dt><var>imagedata</var> = <var>context</var> . <code subdfn data-x="dom-context-2d-getImageData">getImageData</code>(<var>sx</var>, <var>sy</var>, <var>sw</var>, <var>sh</var>)</dt>
+   <dt><var>imagedata</var> = <var>context</var> . <code subdfn data-x="dom-context-2d-createImageData-from-data">createImageData</code>(<var>imagedata</var>)</dt>
+   <dd><p>Returns an <code>ImageData</code> object with the same dimensions and color space as the argument. All the
```
Fixed

ccameron-chromium commented on source:
```diff
@@ -64118,72 +64166,115 @@ try {
-    the actual <code data-x="idl-Uint8ClampedArray">Uint8ClampedArray</code> object passed as the
-    first argument to the constructor.</p>
+    <p class="note">This step does not set <span>this</span>'s data to a copy of <var>data</var>.
+    It's sets it to the actual <code data-x="idl-Uint8ClampedArray">Uint8ClampedArray</code> object
```
I think an intervening change fixed this (I don't see the typo here anymore)

ccameron-chromium commented on source:
```diff
@@ -64118,72 +64166,115 @@ try {
+   <span>"<code>IndexSizeError</code>"</span> <code>DOMException</code>.</p></li>
+
+   <!--ADD-TOPIC:Security-->
+   <li><p>If the <code>CanvasRenderingContext2D</code>'s <span data-x="concept-canvas-origin-clean">
```
Fixed

ccameron-chromium commented on source:
```diff
@@ -64232,6 +64341,9 @@ try {
    <li><p>Initialize <var>value</var>'s <code data-x="dom-imagedata-height">height</code> attribute
    to <var>serialized</var>.[[Height]].</p></li>
+
+   <li><p>Initialize <var>value</var>'s <code data-x="dom-imagedata-colorSpace">colorSpace</code> attribute
```
Fixed

ccameron-chromium commented on source:
```diff
@@ -65748,39 +65865,28 @@ interface <dfn>OffscreenCanvasRenderingContext2D</dfn> {
-  match colors obtained through the <code
-  data-x="dom-context-2d-getImageData">getImageData()</code> method.</p>
+  <p>The <code>canvas</code> APIs provide mechanisms for specifying the color space of the canvas's
+  backing store. The default backing store color space for all canvas APIs is sRGB.</p>
```
Updated all of the references to sRGB in the lines of the patch to point to the definition in color level for. There are still a couple of references outside of the parts touched by the patch, which should be mopped up.

ccameron-chromium commented on source:
```diff
@@ -60678,6 +60704,22 @@ context.fillRect(100,0,50,50); // only this square remains</code></pre>
+  data-x="concept-canvas-color-space">color space</span> of the canvas's backing store.</p>
+
+  <p>The "<dfn enum-value for="PredefinedColorSpace"><code
+  data-x="dom-PredefinedColorSpace-srgb">srgb</code></dfn>" value indicates the sRGB color space.</p>
```
Updated both of these and changed the note below to refer only to conversion functions (as the definitions are now settled in the normative text).

ccameron-chromium commented:
> One thing I would really appreciate (and maybe we should add this to the specification as well) is a concrete example.

I've added an example into the ImageData set of examples in the spec, drawing #FF0000 to a canvas and reading back (234, 51, 35) from getImageData.

annevk commented:
Thanks Christopher! Would you like to list your name in the Acknowledgments section? I will give this a final look later and get it landed.

ccameron-chromium commented:
> Can you make sure to file a WebKit bug and update the OP with it? 

Yes, updated with the WebKit bug.

> Also can we get a commitment to test all the things that changed during review, e.g. handling of invalid values, getImageData-then-putImageData, ... any others I missed?

Yes. We are currently working on adding tests for the behaviors that we have updated during iteration here (the things you mention -- handling of invalid values, and also the default behavior for getImageData and createImageData, once we get them working).

That said, several of the tests have already rolled into upstream (getImageData, toBlob, and putImageData, but all with explicit arguments).

ccameron-chromium commented:
> Thanks Christopher! Would you like to list your name in the Acknowledgments section? I will give this a final look later and get it landed.

Thank you! Yes, certainly, add my name.

annevk commented:
@ccameron-chromium it seems that @domenic's editorial commit got lost when you rebased. I'll try to revive what I can from https://github.com/whatwg/html/compare/eb6c996d7461ac4175b5a47ef381f0fcaa9845ac..93fea05ccf4222977becb51530e85810c24ce218.

annevk commented:
I think all is in order now, but given the mishap I'd like to wait for @domenic to return so he can do a final check and land this Monday.

And once it's landed we can do a small follow-up PR to remove the remaining sRGB references.
